docroot/modules/custom/uiowa_alerts/es6/alerts.es6

```diff
@@ -1,89 +1,129 @@
+      function updateAlerts() {
+
+        // Craft a new `XMLHttpRequest()`.
+        const xhttp = new XMLHttpRequest();
+
+        // Define a function to be called when the readystate changes in the response.
+        xhttp.onreadystatechange = function() {
+
+          // If the readyState is `4`(which in this case means `DONE`),
+          // and we got a 200 range status for a successful response...
+          if (this.readyState === 4 && this.status === 200) {
+            handleResponse(this);
+          }
         };
 
-        // Get alerts on page load.
-        updateAlerts(messages);
+        // Set up the parameters for the request and send it.
+        xhttp.open('GET', settings.uiowaAlerts.source, true);
+        xhttp.send();
+      }
+
+      // Handle the response gotten from the alerts endpoint API call.
+      function handleResponse(response) {
+
+        let newAlerts = [];
+
+        // Parse the JSON string into a usable array.
+        const responseJSON = JSON.parse(response.responseText);
+
+        // For each item we find in the JSON data...
+        responseJSON.data.forEach((item, i) => {
+
+          // Create a new ID for it using the UNIX date to avoid collision
+          // and add it to our new alerts
+          const id = `hawk-alert-${item.attributes.date}`;
+          newAlerts.push(id);
+
+          // If it is already in the Drupal messages section, do not continue.
+          if (messages.select(id)) {
+            return;
+          }
+
+          // Get the crafted alert markup
+          const alert = alertMarkup(item);
+
+          // And add it to the Drupal messages section with its unique ID and message type.
+          messages.add(alert, {
+            id: id,
+            type: 'warning'
+          });
+
+          // Look for differences in existing and new alerts and remove any closed alerts.
+          const difference = existingAlerts.filter(existingAlert => !newAlerts.includes(existingAlert));
+          difference.forEach((closed) => {
+            messages.remove(closed);
+          })
+
+          //Then set existing alerts to the new alerts for the next cycle.
+          existingAlerts = newAlerts;
+        })
+      }
+
+      // Gets existing alerts tracked in the messages wrapper.
+      // Returns an array of DOM elements.
+      function getExistingAlerts() {
+        const existing = [];
+        messagesWrapper.querySelectorAll('.messages').forEach( (existingAlert) => {
+          existing.push(existingAlert.getAttribute('data-drupal-message-id'));
+        });
+
+        return existing;
+      }
+
+      // Takes a JSON item and creates the markup for an alert.
+      // Returns a string of HTML.
+      function alertMarkup(responseJSONItem) {
+        const item = responseJSONItem;
+        const date = new Date(item.attributes.date); // parse the ISO 8601 timestamp
+
+        // Create DateTimeFormat instances with the options
```

Comment:
late to the review party, but should we move these `new Date(...`, `new Intl.DateTimeFormat(...` out so that they are only created once and not for each alert (assuming there are more than one)?

haven't pulled/tested but this caught my eye.

Reply:
oops. i didn't see it got merged. i added this comment from the files tab and didn't see the pr activity. maybe for a later optimization.

Reply:
Good idea. I can fix that if I get the chance